### PR TITLE
flux/coadd

### PR DESCRIPTION
**Pull request by feigewang**
echelle flux/coadd and coadding QA.

### CODE REVIEW

File: pypeit/core/coadd.py [hunk 1]
```diff
@@ -228,7 +230,8 @@ def sn_weight(spectra, smask, debug=False):
 
     rms_sn = np.sqrt(np.mean(sn2)) # Root Mean S/N**2 value for all spectra
 
-    if rms_sn <= 4.0:
+    # FW: should we always using wavelength dependent weights for coadding. at least for megering orders.
```
Comment (profxj) [inline pypeit/core/coadd.py]: Wavelength dependent weights are tricky.  There can be large
variations.  Has always been safest to do order-by-order or 
spectrum-by-spectrum weighting.

File: pypeit/core/coadd.py [hunk 2]
```diff
@@ -1027,50 +1055,70 @@ def coaddspec_qa(ispectra, rspec, rmask, spec1d, qafile=None, yscale=10.):
...
+    ax2.set_xlim([np.min(spec1d.wavelength.value),np.max(spec1d.wavelength.value)])
+    ax2.set_ylim(ylim)
+    ax1.set_xlabel('Wavelength (Angstrom)')
+    ax1.set_ylabel('Flux')
+    ax2.set_ylabel('Flux')
 
     plt.tight_layout(pad=0.2,h_pad=0.,w_pad=0.2)
     if qafile is not None:
-        pp.savefig(bbox_inches='tight')
+        pp.savefig(bbox_inches='tight',dpi=600)
```
[the remark this answers is not in the document]
Comment (feigewang) [reply pypeit/core/coadd.py]: Thanks. I removed dpi=600. It called by coadd_1dspec with vectorized pdf.

File: pypeit/core/coadd.py (hunk 1 above)
Comment (feigewang) [reply pypeit/core/coadd.py]: okay. removed the comments. 
For Echelle, we have options to coadd order-by-order (default) or coadd everything together.

File: pypeit/ech_coadd.py [hunk 3]
```diff
@@ -0,0 +1,201 @@
+
```
Comment (profxj) [inline pypeit/ech_coadd.py]: Add a comment describing the purpose of this module.

File: pypeit/ech_coadd.py [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+from pypeit.core import coadd
+from pypeit.core import load
+from pypeit import msgs
+
+from linetools.spectra.utils import collate
+from linetools.spectra.xspectrum1d import XSpectrum1D
+
+## ToDo: change it to a CLASS and modify coadd_1dspec.py
+def load_spec_order(fname,objid=None,order=None,extract='OPT',flux=True):
```
Comment (profxj) [inline pypeit/ech_coadd.py]: This should probably be in core/load.py, right?

File: pypeit/ech_coadd.py [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+    wave_out,flux_out,sig_out = spectrum.wavelength[~bad_all],spectrum.flux[~bad_all],spectrum.sig[~bad_all]
+    spectrum_out = XSpectrum1D.from_tuple((wave_out,flux_out,sig_out), verbose=False)
+    #if np.sum(bad_flux):
+    #    msgs.warn("There are some bad flux values in this spectrum.  Will zero them out and mask them (not ideal)")
+    #    spectrum.data['flux'][spectrum.select][bad_flux] = 0.
+    #    spectrum.data['sig'][spectrum.select][bad_flux] = 0.
+
+    return spectrum_out
+
+def ech_load_spec(files,objid=None,order=None,extract='OPT',flux=True):
```
Comment (profxj) [inline pypeit/ech_coadd.py]: Same as above.

File: pypeit/ech_coadd.py [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+                       spectrum.sig ** 2 > 1e10,
+                       ], axis=0)
+    if np.sum(bad_flux):
+        msgs.warn("There are some bad flux values in this spectrum.  Will zero them out and mask them (not ideal)")
+        spectrum.data['flux'][spectrum.select][bad_flux] = 0.
+        spectrum.data['sig'][spectrum.select][bad_flux] = 0.
+    return spectrum
+
+
+def ech_coadd(files,objids=None,extract='OPT',flux=True,giantcoadd=False,
```
Comment (profxj) [inline pypeit/ech_coadd.py]: Need a doc string

File: pypeit/fluxspec.py [hunk 7]
```diff
@@ -555,3 +558,390 @@ def __repr__(self):
         txt += '>'
         return txt
 
+
+class EchFluxSpec(masterframe.MasterFrame):
```
Comment (profxj) [inline pypeit/fluxspec.py]: It is unfortunate that we need an entirely separate class
but I'll admit I don't see an easy way to merge the two.

File: pypeit/fluxspec.py [hunk 8]
```diff
@@ -555,3 +558,390 @@ def __repr__(self):
...
+        self.std_idx = None  # Nested indices for the std_specobjs list that corresponds
+        # to the star!
+        # Echelle key
+        self.star_type = star_type
+        self.star_mag = star_mag
+        self.BALM_MASK_WID = BALM_MASK_WID
+        self.nresln = nresln
+        self.debug = debug
+
+    def load_master(self, filename, force=False):
```
Comment (profxj) [inline pypeit/fluxspec.py]: doc string

File: pypeit/fluxspec.py [hunk 9]
```diff
@@ -555,3 +558,390 @@ def __repr__(self):
...
+                            'std_name', 'cal_file', 'ech_orderindx']:
+                    try:
+                        sens_dict[key] = head[key.upper()]
+                    except:
+                        pass
+                sens_dicts[str(iord)] = sens_dict
+            sens_dicts['norder'] = norder
+            return sens_dicts
+
+    def save_master(self, sens_dicts, outfile=None):
```
Comment (profxj) [inline pypeit/fluxspec.py]: define sens_dicts

File: pypeit/fluxspec.py [hunk 10]
```diff
@@ -555,3 +558,390 @@ def __repr__(self):
...
+        self.sci_spec1d_file = sci_spec1d_file
+        self.sci_specobjs = []
+        self.sci_header = None
+        if self.sci_spec1d_file is not None:
+            self.sci_specobjs, self.sci_header = load.ech_load_specobj(self.sci_spec1d_file)
+            msgs.info('Loaded {0} spectra from the spec1d science file: {1}'.format(
+                len(self.sci_specobjs), self.sci_spec1d_file))
+            sci_spectro = self.sci_header['INSTRUME']
+
+        # Compare instruments if they exist
```
Comment (jhennawi) [inline pypeit/fluxspec.py]: I'm lost here, when will we ever be fluxing from a different instrument?

File: pypeit/core/load.py [hunk 11]
```diff
@@ -144,6 +144,87 @@ def load_specobj(fname):
     return specObjs, head0
 
 
+def ech_load_specobj(fname, order=None):
+
+    """ Load a spec1d file into a list of SpecObjExp objects
```
[the remark this answers is not in the document]
Comment (feigewang) [reply pypeit/core/load.py]: done

File: pypeit/core/load.py [hunk 12]
```diff
@@ -144,6 +144,87 @@ def load_specobj(fname):
...
+    head0 = hdulist[0].header
+    for hdu in hdulist:
+        if hdu.name == 'PRIMARY':
+            continue
+        # elif hdu.name[8:17] != 'ORDER'+'{0:04}'.format(order):
+        #    continue
+        # Parse name
+        idx = hdu.name
+        objp = idx.split('-')
+        if objp[-2][0:3] == 'DET':
```
[the remark this answers is not in the document]
Comment (feigewang) [reply pypeit/core/load.py]: yes. it is a bit tricky here. we may want define an Echelle model somewhere else. I add ToDo here, will back to this after we finalize names for Echelle data model.

File: pypeit/ech_coadd.py (hunk 3 above)
Comment (feigewang) [reply pypeit/ech_coadd.py]: done

File: pypeit/ech_coadd.py (hunk 4 above)
Comment (feigewang) [reply pypeit/ech_coadd.py]: done

File: pypeit/ech_coadd.py (hunk 5 above)
Comment (feigewang) [reply pypeit/ech_coadd.py]: done

File: pypeit/ech_coadd.py (hunk 6 above)
Comment (feigewang) [reply pypeit/ech_coadd.py]: done

File: pypeit/fluxspec.py (hunk 7 above)
Comment (feigewang) [reply pypeit/fluxspec.py]: I also didn't find a good way to merge these two classes. It's easy to be called by the flux script to make is a master class.

File: pypeit/fluxspec.py (hunk 8 above)
Comment (feigewang) [reply pypeit/fluxspec.py]: done

File: pypeit/fluxspec.py (hunk 9 above)
Comment (feigewang) [reply pypeit/fluxspec.py]: done

File: pypeit/core/coadd.py [hunk 13]
```diff
@@ -228,7 +230,7 @@ def sn_weight(spectra, smask, debug=False):
 
     rms_sn = np.sqrt(np.mean(sn2)) # Root Mean S/N**2 value for all spectra
 
-    if rms_sn <= 4.0:
+    if rms_sn <= 3.0:
```
[the remark this answers is not in the document]
Comment (feigewang) [reply pypeit/core/coadd.py]: Add a todo here.

File: pypeit/fluxspec.py (hunk 10 above)
Comment (feigewang) [reply pypeit/fluxspec.py]: we used the coordinates of the telescope to correct the extinction.

**Comment by profxj** [comment]
Approving and merging.

**Pull request merged**